codeface/R/cluster/graph_comparison.r

```diff
@@ -142,37 +173,134 @@ graph.comparison <- function(g.1, g.2) {
+  local.ids.2 <- map.ids(global.ids.2, map2)
+  edgelist.local.2 <- data.frame(from=map.ids(edgelist.2$fromId, map2),
+                                 to=map.ids(edgelist.2$toId, map2),
+                                 weight=edgelist.2$weight)
+  vertex.df.2 <- data.frame(id2=local.ids.2, name=node.label.2)
+
+  directed <- T
+  ## create graph instances and run the graph comparison.
+
+  # A graph with nno weights and only the local vertices
+  unmerged.unweighted.g.1 <- graph.data.frame(
+    data.frame(from=edgelist.local.1$from, to=edgelist.local.1$to),
+    vertices=vertex.df.1, directed=directed)
+  unmerged.unweighted.g.2 <- graph.data.frame(
+    data.frame(from=edgelist.local.2$from, to=edgelist.local.2$to),
+    vertices=vertex.df.2, directed=directed)
+  V(unmerged.unweighted.g.1)$Id <- vertex.df.1$name
+  V(unmerged.unweighted.g.2)$Id <- vertex.df.2$name
+
+  # A graph with weights and all vertices
+  g.1 <- graph.data.frame(edgelist.local.1, vertices=merged.vertex, directed=directed)
+  g.2 <- graph.data.frame(edgelist.local.2, vertices=merged.vertex, directed=directed)
+  V(g.1)$Id <- merged.vertex$name
+  V(g.2)$Id <- merged.vertex$name
+
+  # vertex.diff describes how different the edges are, we now calculate how different the graphs are
+  vertex.diff <- graph.comparison(g.1, g.2, weighted = weighted, symmetric = symmetric)
+
+  intersectIds <- intersect(local.ids.1, local.ids.2)
+  nodes.diff <- 1 - (length(intersectIds) / nrow(merged.vertex))
+
+  # Only works when we use the merged.vertex graphs
+  vertexList <- V(g.1)
+  total.weight <- 0
+  total.weighted.diff <- 0
+  for (v in vertexList) {
+    if (v %in% intersectIds) {
+      weight <- sum(c(edgelist.local.1$weight[v], edgelist.local.2$weight[v]), na.rm=T)
+      total.weighted.diff <- total.weighted.diff + (vertex.diff[v] * weight)
+      total.weight <- total.weight + weight
+    }
+  }
+
+  vertex.weighted.diff <- total.weighted.diff / total.weight
+
+  total <- 0
+  total.diff <- 0
+  for (v in vertexList) {
+    if (v %in% intersectIds) {
+      total.diff <- total.diff + vertex.diff[v]
+      total <- total + 1
+    }
+  }
+  vertex.total.diff <- total.diff / total
 
-  res <- graph.comparison(g.1, g.2)
-  return(res)
+  return (list(vertex.diff=vertex.diff, nodes.diff=nodes.diff,
+               vertex.weighted.diff=vertex.weighted.diff,
+               vertex.total.diff=vertex.total.diff,
+               cohesion.1 = graph.cohesion(unmerged.unweighted.g.1),
+               cohesion.2 = graph.cohesion(unmerged.unweighted.g.2),
+               diameter.1 = diameter(unmerged.unweighted.g.1),
+               diameter.2 = diameter(unmerged.unweighted.g.2),
+               density.1 = graph.density(unmerged.unweighted.g.1),
+               density.2 = graph.density(unmerged.unweighted.g.2),
+               transitivity.1 = transitivity(unmerged.unweighted.g.1),
+               transitivity.2 = transitivity(unmerged.unweighted.g.2)))
```

Comment:
We have a lot of graph analysis code in another file but this is fine to give some context for now.